core-authentication/src/main/java/gg/solarmc/loader/authentication/AuthenticationCenter.java

```diff
@@ -69,59 +151,67 @@ public InitialLoginAttempt attemptInitialLogin(Transaction transaction, Unauthen
 				.fetchOne();
 		if (passwordRecord == null) {
 			if (user.isPremium()) {
-				// New premium user, insert empty password and permit login
+				// New premium username, insert empty password and permit login
 				insertAutomaticAccount(transaction, user);
-				return InitialLoginAttempt.premiumPermitted();
+				return LoginAttempt.premiumPermitted();
 			}
-			// New cracked user, request account creation
-			return InitialLoginAttempt.needsAccount();
+			// New cracked username, request account creation
+			return LoginAttempt.needsAccount();
 		}
+		LoggerFactory.getLogger(getClass()).info(
+				"An incredible race condition occurred (or this is a bug). User {} logged in " +
+				"while no one owned their username. Now, however, someone has taken their username.",
+				user.username());
 		int iterations = passwordRecord.get(AUTH_PASSWORDS.ITERATIONS);
 		if (iterations == 0) { // 0 indicates a premium account
 			/*
 			Existing premium/automatic account
 			 */
 			if (!user.isPremium()) {
 				// User is cracked, but is trying to use a premium user's name
-				return InitialLoginAttempt.deniedPremiumTookName();
+				return LoginAttempt.deniedPremiumTookName();
 			}
-			// User must be permitted. This has to be the same premium user
+			/*
+			 * The same premium user is involved with a race condition among themselves?
+			 * What sorcery is this? Multiple people on the same account?
+			 * In any case, we can actually permit this user to join. They may see
+			 * outdated cached values for their user stats if they play at the same time;
+			 * however, this is no concern for the reliability of their actual data
+			 * stored in the database. We guarantee complete data reliability.
+			 */
 			int userId = fetchIdFromUuid(context, user);
 			user.loadData(transaction, userId);
-			return InitialLoginAttempt.premiumPermitted();
+			return LoginAttempt.premiumPermitted();
 		}
 		// Existing cracked account
-
 		if (!user.username().equals(passwordRecord.get(AUTH_PASSWORDS.USERNAME))) {
-			/*
-			MariaDB is case-insensitive, which is why we were able to select this account.
-			However, we cannot permit multiple cracked users to have the same name ignoring case.
```

Comment:
Are minecraft names case sensitive? What about cracked user names, can someone set their name to someOne if sOmeone already exists?

Reply:
Names are case-insensitive for the most part, but there is a catch with respect to cracked users.

Usernames of normal accounts (premium accounts) are case-insensitive in every way. There cannot be 2 users with the same name ignoring case.

Usernames of cracked accounts have odd behavior if adequate protections are not taken. 2 users with the same name ignoring case, but a different name including case, will have different UUIDs. This is one too many of an edge case to handle safely, and it will only cause issues. Some parts of the server operate on the premise of case-insensitivity of names. Our name/UUID records also assume case-insensitivity of names. I'm sure other plugins do too. Working around this principle of case-insensitive names (which is a perfectly reasonable principle for premium players, but not for this particular case of cracked players) adds unnecessary difficulty to an already complicated situation, when we can just disallow this very rare case and avoid all that work.